src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -1807,36 +1845,38 @@ public override IOperation VisitInvocation(IInvocationOperation operation, int?
             }
 
             ImmutableArray<IArgumentOperation> arguments = operation.Arguments;
-            foreach (IArgumentOperation argument in arguments)
-            {
-                _evalStack.Push(Visit(argument.Value));
-            }
-
-            ImmutableArray<IArgumentOperation> visitedArguments;
-            if (!arguments.IsEmpty)
-            {
-                var builder = ArrayBuilder<IArgumentOperation>.GetInstance();
-                for (int i = arguments.Length - 1; i >= 0; i--)
-                {
-                    IOperation visitedValue = _evalStack.Pop();
-                    var originalArgument = (BaseArgument)arguments[i];
-                    builder.Add(new ArgumentOperation(visitedValue, originalArgument.ArgumentKind, originalArgument.Parameter,
-                                                      originalArgument.InConversionConvertibleOpt, originalArgument.OutConversionConvertibleOpt,
-                                                      semanticModel: null, originalArgument.Syntax, originalArgument.ConstantValue, originalArgument.IsImplicit));
-                }
-                builder.ReverseContents();
-                visitedArguments = builder.ToImmutableAndFree();
-            }
-            else
-            {
-                visitedArguments = ImmutableArray<IArgumentOperation>.Empty;
-            }
-
+            PushArray(arguments, arg => arg.Value);
+            ImmutableArray<IArgumentOperation> visitedArguments = PopArray(arguments, _argumentRewriter);
```

Comment:
Consider extracting a `VisitArguments(operation.Arguments)` method. That might be useful for a number of places: invocation, object creation, indexer, object initializer member, etc., and perhaps `PushArray` and `PopArray` could be inlined in that method.

Reply:
I considered this. However, Push/PopArray were designed to used for any `ImmutableArray<T> where T : IOperation`, and that ImmutableArray isn't always the last thing that needs to be visited. So Push needs to happen, then other things need to be visited, and then we can Pop things off the stack. Looking at IArgumentOperation's usage, though, it looks like we will always want to Push/Pop last, so that can be factored into a helper.

---
In reply to: [172571473](https://github.com/dotnet/roslyn/pull/25178#discussion_r172571473) [](ancestors = 172571473)

Reply:
I'll add a helper for arguments and a prototype comment to revisit whether these are necessary after we're done implementation. I'm of the opinion they'll be useful, but we can remove later if they're not.

---
In reply to: [172669449](https://github.com/dotnet/roslyn/pull/25178#discussion_r172669449) [](ancestors = 172669449,172571473)